ldap/servers/slapd/daemon.c

```diff
@@ -1778,6 +1789,99 @@ daemon_register_connection()
 }
 
 #if defined(ENABLE_LDAPI)
+struct ldapi_mapping {
+    char *dn;
+    uid_t uid;
+    gid_t gid;
+    struct ldapi_mapping *next;
+};
+
+static struct ldapi_mapping *ldapi_mappings = NULL;
+
+static void
+initialize_ldapi_auth_dn_mappings(void)
+{
+    Slapi_PBlock *search_pb = NULL;
+    Slapi_Entry **entries = NULL;
+    struct ldapi_mapping *mappings = NULL;
+    char *base_dn = config_get_ldapi_mapping_base_dn();
+    char *filter = "(|(objectclass=nsLDAPIAuthMap)(objectclass=nsLDAPIFixedAuthMap))";
+    int result = 0;
+
+    if (base_dn == NULL) {
+        /* nothing to do */
+        return;
+    }
+
+    /* Get all the mapping entries */
+    search_pb = slapi_pblock_new();
+    slapi_search_internal_set_pb(search_pb, base_dn, LDAP_SCOPE_SUBTREE, filter,
+            NULL, 0, NULL, NULL, (void *)plugin_get_default_component_id(), 0);
+    slapi_search_internal_pb(search_pb);
+    slapi_pblock_get(search_pb, SLAPI_PLUGIN_INTOP_RESULT, &result);
+    if (result == LDAP_SUCCESS) {
+        slapi_pblock_get(search_pb, SLAPI_PLUGIN_INTOP_SEARCH_ENTRIES, &entries);
+        for (size_t i = 0; entries && entries[i]; i++) {
+            const char *username = slapi_entry_attr_get_ref(entries[i], CONFIG_LDAPI_AUTH_USERNAME_ATTRIBUTE);
+            uid_t user_uid = slapi_entry_attr_get_uint(entries[i], "uidNumber");
+            gid_t user_gid = slapi_entry_attr_get_uint(entries[i], "gidNumber");
+
+            if (user_uid && user_gid) {
```

Comment:
If we are using this logic, perhaps it would be worth it being a single objectClass and we require the uid to be set, but then you can force-override a uid/gid number if you want?

Reply:
This could be improved, because I wrote the schema to enforce that the needed values are set.  Either uid & gid are both NULL or they are both set (we don't need to check both are not NULL).  It's one or the other, and the code does not represent that very well.  So I'd prefer to keep the two objectclasses, and let me revise this code a bit...

Reply:
That's fine, I'm not too stubborn about this detail.